task/net/src/server.rs

```diff
@@ -559,9 +567,30 @@ where
                     .read_range(0..payload.len(), buf)
                     .map_err(|_| RequestError::went_away())?;
                 self.client_waiting_to_send[socket_index] = false;
+                self.first_queue_full[socket_index] = None;
                 Ok(())
             }
             Err(smoltcp::Error::Exhausted) => {
+                // Workaround for a smoltcp issue:
+                // https://github.com/smoltcp-rs/smoltcp/issues/594
+                //
+                // If we've gotten nothing but `QueueFull` errors for a certain
+                // amount of time, assume that the queue is clogged and panic
+                // the task.
+                //
+                // Note that any successful packet send on this socket will
+                // reset `self.first_queue_full`, because it means that the
+                // queue isn't clogged.
+                let now = userlib::sys_get_timer().now;
+                if let Some(t) = self.first_queue_full[socket_index] {
+                    // Leave `self.first_queue_full` unchanged; it's already
+                    // populated with the time of the first QueueFull error.
+                    if now >= t + 500 {
```

Comment:
Maaaaybe worth putting `500` into a named constant that's greppable? Since it's only used once it's not a very strong argument though.